### DISCUSSION TITLE
Dynamic response model: How to allow request to alter response_model_include values?

### DISCUSSION THREAD

**Discussion by alukach** [fastapi/fastapi#3310] [Questions] [answered]
Is there any way to do something like `response_model_include` but have it by dynamically set via query parameters?  I would like to support the user filtering output fields by a field query parameter.

Trying to do something like the following fails because when I filter within my request handler, the response object no longer complies with my Pydantic response model:

```py
from pydantic import BaseModel

class Character(BaseModel):
    name: str
    age: int
    height: int


@app.get("/", response_model=Character)
async def example(fields: Set[str] = Query([])):
    results = {"name": "Donald Duck", "age": 34, "height": 172}
    if fields:
        return {k: v for k, v in results.items() if k in fields}
    return results
```

Am I on the right track but I need to mark every field in the `Character` class as `Optional[...]`?  Given the wealth of FastAPI's convenience features, I feel like there must be a cleaner way to do this but have failed to come across it after combing through the [ResponseModel](https://fastapi.tiangolo.com/tutorial/response-model/) docs.

**Comment by nzig** [comment]
I don't think that should be possible. Doing so would make the OpenAPI schema wrong, and would also violate typing declarations by having `None` in non-`Optional` fields.

**Comment by alukach** [comment, marked as answer]
In the end, I settled on wrapping every property with `Optional[]`, setting them to `None` by default, and utilizing the `response_model_exclude_unset` config:

```py
from typings import Optional, Set

from fastapi import Query
from pydantic import BaseModel

class Character(BaseModel):
    name: Optional[str] = None
    age: Optional[int] = None
    height: Optional[int] = None


@app.get("/", response_model=Character, response_model_exclude_unset=True)
async def example(fields: Set[str] = Query([])):
    results = {"name": "Donald Duck", "age": 34, "height": 172}
    if fields:
        return {k: v for k, v in results.items() if k in fields}
    return results
```

This provides the user experience I am looking for while also being true to the OpenAPI schema.
